### PR TITLE
Preview package

### PR DESCRIPTION
**Pull request by KevinRansom**
1.  Enable the netsdk to work with preview versions of FSharp.Core nuget packages.

Add to the project file this property:
````
       <UseFSharpPreview>true</UseFSharpPreview>
````
The build will pick up the latest preview file published to myget.

2.  Parameterize, creation of foo.exe.config in repo build to set fsharp.core binding redirects.
3.  Parameterize F# Banner text.

### CODE REVIEW

File: FSharpBuild.Directory.Build.targets [hunk 1]
```diff
@@ -4,23 +4,50 @@
...
+
+    <WriteLinesToFile File="@(CopyAndSubstituteText->'$(OutDir)%(Filename)%(Extension)')" Lines="$(_ReplacementText)" Overwrite="true" />
+
     <!-- Make sure it will get cleaned  -->
-    <CreateItem Include="$(OutDir)%(CopyAndSubstituteText.TargetFilename)">
+    <ItemGroup>
+        <None Include="@(CopyAndSubstituteText->'$(OutDir)%(Filename)%(Extension)')" CopyToOutputDirectory="PreserveNewest" />
+    </ItemGroup>
+
+    <CreateItem Include="@(CopyAndSubstituteText->'$(OutDir)%(Filename)%(Extension)')">
```
Comment (brettfo) [inline FSharpBuild.Directory.Build.targets]: Do we still need this?  `<WriteLinesToFile>` should ensure it exists on disk and `<None Include...>` should make sure it ends up in the `<FileWrites>` item group.

File: FSharpBuild.Directory.Build.targets [hunk 2]
```diff
@@ -4,23 +4,50 @@
...
+      <_BuildPropertyLines Include="module internal FSharp.BuildProperties" />
+      <_BuildPropertyLines Include="let fsProductVersion = &quot;$(FSPRODUCTVERSION)&quot;" />
+      <_BuildPropertyLines Include="let fsLanguageVersion = &quot;$(FSLANGUAGEVERSION)&quot;" />
+    </ItemGroup>
+
+    <WriteLinesToFile File="$(IntermediateOutputPath)buildproperties.fs" Lines="@(_BuildPropertyLines)" Overwrite="true" />
+
+    <!-- Make sure it will get cleaned  -->
+    <ItemGroup>
+      <CompileBefore Include="$(IntermediateOutputPath)buildproperties.fs" CopyToOutputDirectory="PreserveNewest" />
```
Comment (brettfo) [inline FSharpBuild.Directory.Build.targets]: Also add to `<FileWrites>`.

File: eng/Versions.props [hunk 3]
```diff
@@ -9,27 +9,34 @@
     <UsingToolSymbolUploader>true</UsingToolSymbolUploader>
     <UsingToolVSSDK>true</UsingToolVSSDK>
   </PropertyGroup>
+
```
Comment (brettfo) [inline eng/Versions.props]: Nit: The Arcade update bot is going to remove all empty lines the next time it comes through.

File: eng/Versions.props [hunk 4]
```diff
@@ -9,27 +9,34 @@
...
   <!-- Version number computation -->
   <PropertyGroup>
     <PreReleaseVersionLabel>beta</PreReleaseVersionLabel>
-    <FSCoreMajorVersion>4.6</FSCoreMajorVersion>
+    <FSLanguageVersion>4.6</FSLanguageVersion>
+    <FSCoreMajorVersion>$(FSLanguageVersion)</FSCoreMajorVersion>
     <FSCorePackageVersion>$(FSCoreMajorVersion).3</FSCorePackageVersion>
     <FSCoreVersionPrefix>$(FSCoreMajorVersion).0</FSCoreVersionPrefix>
     <FSCoreVersion>$(FSCoreVersionPrefix).0</FSCoreVersion>
+    <FSharpCoreShippedPackageVersion>4.6.2</FSharpCoreShippedPackageVersion>                                <!-- The current published nuget package -->
```
Comment (brettfo) [inline eng/Versions.props]: The arcade update bot will also screw up the formatting of these.  To keep them relevant add them on the line before.

File: src/fsharp/FSharp.Build/Microsoft.FSharp.NetSdk.targets [hunk 5]
```diff
@@ -52,11 +52,17 @@ WARNING:  DO NOT MODIFY this file unless you are knowledgeable about MSBuild and
...
+    <FSharpCoreImplicitPackageVersion>$(DefaultFSharpCorePreviewPackageVersion)</FSharpCoreImplicitPackageVersion>
+    <RestoreSources>$(RestoreSources); https://dotnet.myget.org/F/fsharp/api/v3/index.json</RestoreSources>
+  </PropertyGroup>
+
   <Target Name="PackageFSharpDesignTimeTools" DependsOnTargets="_GetFrameworkAssemblyReferences">
     <PropertyGroup>
       <FSharpDesignTimeProtocol Condition = " '$(FSharpDesignTimeProtocol)' == '' ">fsharp41</FSharpDesignTimeProtocol>
       <FSharpToolsDirectory Condition = " '$(FSharpToolsDirectory)' == '' ">tools</FSharpToolsDirectory>
-    </PropertyGroup>
+    </PropertyGroup>
```
Comment (brettfo) [inline src/fsharp/FSharp.Build/Microsoft.FSharp.NetSdk.targets]: Nit: trailing whitespace.

File: src/fsharp/FSharp.Build/xlf/FSBuild.txt.cs.xlf [hunk 6]
```diff
@@ -2,6 +2,11 @@
 <xliff xmlns="urn:oasis:names:tc:xliff:document:1.2" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" version="1.2" xsi:schemaLocation="urn:oasis:names:tc:xliff:document:1.2 xliff-core-1.2-transitional.xsd">
   <file datatype="xml" source-language="en" target-language="cs" original="../FSBuild.resx">
     <body>
+      <trans-unit id="fSharpBannerVersion">
+        <source>{0} for F# {1}</source>
+        <target state="new">{0} for F# {1}</target>
+        <note />
```
Comment (brettfo) [inline src/fsharp/FSharp.Build/xlf/FSBuild.txt.cs.xlf]: Might want to change this to: `<note>Contains version numbers.  E.g., '10.4.3 for F# 4.6'</note>` so the translators know what's happening.

Comment (KevinRansom) [reply src/fsharp/FSharp.Build/xlf/FSBuild.txt.cs.xlf]: I like that … good idea.

Comment (KevinRansom) [reply src/fsharp/FSharp.Build/xlf/FSBuild.txt.cs.xlf]: I tried it, and the note I had added got overwritten.

**Pull request merged**